core/shared/src/main/scala/scalaz/zio/Duration.scala

```diff
@@ -1,25 +1,29 @@
 package scalaz.zio
 
-sealed trait Duration {
-  def toDuration: scala.concurrent.duration.Duration
-}
+sealed trait Duration
 
 object Duration {
 
-  case class Finite(duration: scala.concurrent.duration.FiniteDuration) extends Duration {
-    final def toDuration = duration
-  }
+  class Finite(duration: scala.concurrent.duration.FiniteDuration) extends Duration {
```

Comment:
Make it:

```scala
case class Finite(private val duration: scala.concurrent.duration.FiniteDuration) extends Duration {
```

Same effect but you get the equality/hashCode/etc. from the case class.

Reply:
Actually, no, let's change it to this:

```scala
case class Finite(length: Long, unit: TimeUnit) {
}
```

This way users can deconstruct directly on `Finite`, e.g.:

```scala
d match {
  case Finite(length, unit) => 
  case Infinite =>
}
```

Reply:
@jdegoes This will require implementing `equals` and `hashCode` based on `toNanos` which partially diminishes the return of using a `case class`. And I actually wonder if such a deconstruction of `length` and `unit` is even useful. These values don't seem that interesting after creating the `Duration`. `toNanos` and `toMillis` seem to be the way to go pretty much every time.

Reply:
How about this, then:

```scala
case class Finite(toNanos: Long)
```

Reply:
I guess it could work. So we always go down to nanoseconds when creating a `Duration`.
This way we can drop the underlying `scala.concurrent.duration.Duration` after all. I mean it still could give us proven implementations of some of the operations, but now creating it specifically for that doesn't seem that reasonable.

Reply:
@jdegoes ok, one problem, we can't really do much in case `Finite` is created with a negative `Long`

Reply:
Can't win. Maddening! 👹 

We could hide all the constructors and even the types with a single fold method on `Duration`:

```scala
def fold[Z](infinity: => Z, nanos: Long => Z): Z
```

Not sure what that would do to ergonomics, though.

Reply:
@jdegoes how about this. It seems to give us what we've been talking about - deconstruction and safe instance creation. In short:
```scala
sealed trait Duration

final object Duration {

  object Finite {

    def safeFromNanos(nanos: Long): Option[Finite] = 
      if(nanos >= 0) Some(new Finite(nanos))
      else None

    def unapply(finite: Finite): Option[Long] = Some(finite.nanos)
  }

  final class Finite private (val nanos: Long) extends Duration

  final case object Infinity extends Duration

  def fromNanos(nanos: Long): Duration = Finite.safeFromNanos(nanos).getOrElse(Infinity)

}
```